src/library/scala/collection/Iterator.scala

```diff
@@ -391,9 +391,9 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
 
   def indexWhere(p: A => Boolean, from: Int = 0): Int = {
     var i = math.max(from, 0)
-    drop(from)
-    while (hasNext) {
-      if (p(next())) return i
+    val modifiedIterator = drop(from)
+    while (modifiedIterator.hasNext) {
+      if (p(modifiedIterator.next())) return i
       i += 1
     }
     -1
```

Comment:
This doesn't seem be any different, unless there are cases where `drop()` does not consume the iterator?

Reply:
My `drop` modifications returns a new Iterator. In this case, calling `drop` no longer modifies the state of the current iterator calling the `drop` function, which is why we need to use the reference of the newly returned sliced iterator. 
The drawback of this is that this consumes additional space (because of reference to new iterator), but it's standardized across `take`, `slice` and `drop` operations in the sense they return a new iterator instead of modifying the current one.

Reply:
Any combinator on an Iterator invalidates the original object. There is no reason to unnecessarily allocate a new object and pretend the old one is still usable.